### DISCUSSION TITLE
problems with pip installing numpy

### DISCUSSION THREAD

**Discussion by janetcgreen** [spacepy/spacepy#604] [Support and Q&A] [answered]
My issue is with installing spacepy using pip. Using pip also installs numpy version 1.21.4 and leads to a mess of library issues if the existing python libraries were installed with conda. It creates two different versions of numpy. I can't delete the conda version of numpy because then it tries to delete everything that depends on numpy as well. And I can't upgrade the conda version of numpy to match the spacepy version because the conda channels skip from numpy version 1.21.2 to 1.21.5. Maybe there is something simple but I don't see how to make this work with any anaconda installs of python.

**Comment by jtniehof** [comment]
SpacePy doesn't force a version of numpy (other than a very old minimum version), and in theory pip should find the conda version if it's already there and not try to install anything newer. Theory and practice, however...

Can you let me know:

- What OS you're on, and what version?
- What version of conda you're using (`conda --version`)?
- What version of python (`python --version`)?
- Ideally, a full sequence of commands and inputs from creating a new conda environment through the issue showing up?

On Ubuntu 18.04 and conda 4.12.0 (python 3.9, the default), I activate my conda environment and then:
```
conda install numpy
pip install numpy
```
and I get "Requirement already satisfied: numpy in /home/jtniehof/miniconda/envs/foo/lib/python3.9/site-packages (1.21.5)"

If I do `pip install spacepy` I get, as part of the install, "Requirement already satisfied: numpy!=1.15.0,>=1.10 in /home/jtniehof/miniconda/envs/foo/lib/python3.9/site-packages (from spacepy) (1.21.5)"

I can also force-install a different version with conda (`conda install "numpy==1.20.3"`) and will still get "Requirement already satisfied."

**Comment by janetcgreen** [comment]
I'm having trouble with both linux centos and macos Catalina. I thought it
was related to the fact that pip had installed another version of numpy.
But I think I cleared up the multiple installation issue and I'm still
having troubles.

The real problem is that this import statement in irbempy.py 'from .
import irbempylib
as oplib ' crashes on macos with this error  'RuntimeError: module compiled
against API version 0xf but this version of numpy is 0xe'. The same import
statement crashes on centos with a slightly different but still numpy
related error 'ImportError: numpy.core.multiarray failed to import'.

If you have any thoughts on what the issue might be I would appreciate it.
Thank you for your help.

Janet

On Wed, Apr 13, 2022 at 10:19 AM Jon Niehof ***@***.***>
wrote:

> SpacePy doesn't force a version of numpy (other than a very old minimum
> version), and in theory pip should find the conda version if it's already
> there and not try to install anything newer. Theory and practice, however...
>
> Can you let me know:
>
>    - What OS you're on, and what version?
>    - What version of conda you're using (conda --version)?
>    - What version of python (python --version)?
>    - Ideally, a full sequence of commands and inputs from creating a new
>    conda environment through the issue showing up?
>
> On Ubuntu 18.04 and conda 4.12.0, I activate my conda environment and then:
>
> conda install numpy
> pip install numpy
>
> and I get "Requirement already satisfied: numpy in
> /home/jtniehof/miniconda/envs/foo/lib/python3.9/site-packages (1.21.5)"
>
> If I do pip install spacepy I get, as part of the install, "Requirement
> already satisfied: numpy!=1.15.0,>=1.10 in
> /home/jtniehof/miniconda/envs/foo/lib/python3.9/site-packages (from
> spacepy) (1.21.5)"
>
> I can also force-install a different version with conda (conda install
> "numpy==1.20.3") and will still get "Requirement already satisfied."
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/spacepy/spacepy/issues/590#issuecomment-1098243960>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQJQHFTRR3Q26KLZMBUQ4TDVE3XXDANCNFSM5TIULHVA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Space Hazards Applications
720-222-0533

**Comment by jtniehof** [comment]
The API mismatch strongly suggests either there are still two versions of numpy running around or that SpacePy was compiled against a different version. You might want to try:
```
pip uninstall spacepy
pip cache remove spacepy
pip install spacepy
```
to make sure it's compiled fresh.

The CentOS one is weirder, unless there's also an API mismatch lurking in the full error message. Does just importing numpy itself work?

**Reply by jtniehof** [reply]
I'm seeing the API mismatch on the Mac now (Big Sur), with conda being the only numpy installed. What I think is happening is that, due to #517, pip is installing numpy in a separate environment to do the build. That's a newer version of numpy (1.23, in this case) than conda has installed (1.21), so the built module won't import in that environment. This will _only_ happen when installing with `pip` instead of `setup.py`, so there's a workaround. Another is to explicitly do `pip install numpy` before installing spacepy, which is of course a pain.

If I remove the numpy requirement from pyproject.toml, the build fails because pip is still using an isolated environment. Same if I have an empty pyproject.toml. Completely removing pyproject.toml fixes the problem, but that's not going to work with the migration to the New Way Of Installing Things.

I don't think this is Mac specific, it's just a question of what's available in conda vs. pip on any particular platform.

Bottom line: pip makes no attempt to check if the version used to build a package is the same version that it will be installed under, even though it does check that the requirement exists.

Converting this to an issue (#638)

**Comment by drsteve** [comment]
Just to catch up on this...
I have seen the `ImportError: numpy.core.multiarray failed to import` error before, and it is related to building against a different version of numpy. In a separate conversation with @jtniehof it came up that pip will check for dependencies installed by conda, but not vice versa. That makes it really important to make sure that common dependencies are either installed by conda up front, or that _everything_ is done with pip after making the virtual environment.

My guess is that there are still multiple versions of numpy in the environment here because `pip install spacepy` (or some other package that required numpy) was run before there was a conda installed version of numpy. So then subsequent package installs using conda brought up a requirement for numpy and conda grabbed a version that it manages.

I'm going to move this to discussion, but I'd suggest testing in a clean virtual environment to avoid issues with a messy virtual environment. I'd also suggest checking whether there are other concerns like having things set in your `PYTHONPATH` (not generally recommended when working with virtual environments), as that could be leading your python to find incompatible libraries.
That is:
```
conda create -n testenv -c conda-forge python numpy scipy matplotlib
conda activate testenv
pip install spacepy  # this should find everything it needs except h5py
```
That should install cleanly (I just double checked on Ubuntu 20.04/WSL2) and all the modules should import (I checked `time`, `coordinates`, and `irbempy`, as well as doing some basic function calls).

**Comment by jtniehof** [comment, marked as answer]
I'm going to summarize and close this discussion, because there are several related things that have been addressed in the most recent releases of SpacePy (0.4.0 and 0.4.1):

- pip builds against a different numpy than is installed, #638 fixed in #639 as part of 0.4.0 (and #649 for Apple Silicon, 0.4.1)
- irbem failure to compile, which should be addressed with Mac SDK issues in #639. The updated information on [Mac installation](https://spacepy.github.io/install_mac.html) helps, and there is new documentation on [troubleshooting irbempy issues](https://spacepy.github.io/install.html#irbempy) (all included in 0.4.0)

These are the most likely root causes of the issues in this discussion; anybody running into issues with the latest SpacePy is invited to open a new discussion for help.
